INT-4208: Support for AsycRestTemplate and thus noblocking IO

pkuashi opened the pull request:
https://jira.spring.io/browse/INT-4208

Add AsyncRestTemplate support in HttpRequestExecutingMessageHandler and corresponding config/dsl support

pkuashi-official committed:
INT-4208: Support for AsycRestTemplate and thus noblocking IO

https://jira.spring.io/browse/INT-4208

Add AsyncRestTemplate support in HttpRequestExecutingMessageHandler and corresponding config/dsl support

artembilan commented on spring-integration-http/src/main/java/org/springframework/integration/http/outbound/AsyncHttpRequestExecutingMessageHandler.java:
```diff
@@ -0,0 +1,161 @@
+ * @since 5.0
+ */
+public class AsyncHttpRequestExecutingMessageHandler extends AbstractHttpRequestExecutingMessageHandler {
+	private AsyncRestTemplate asyncRestTemplate;
```
Must be `final` and surrounded with blank lines.

artembilan commented on src/reference/asciidoc/http.adoc:
```diff
@@ -190,6 +191,36 @@ The `expected-response-type` must be compatible with the (configured or default)
+</bean>
+----
+
+You can configure the AsyncClientHttpRequestFactory instance to use:
```
Consider to wrap all the classes mentioned to the code using "`" symbol.

artembilan commented:
Please, also mention in the commit message https://jira.spring.io/browse/INT-4076

pkuashi-official committed:
INT-4208: Support for AsycRestTemplate and thus noblocking IO

https://jira.spring.io/browse/INT-4076
https://jira.spring.io/browse/INT-4208

Add AsyncRestTemplate support in HttpRequestExecutingMessageHandler and corresponding config/dsl support

pkuashi commented:
Will clear the errors tomorrow

pkuashi-official committed:
INT-4208: Support for AsycRestTemplate and thus noblocking IO

https://jira.spring.io/browse/INT-4076
https://jira.spring.io/browse/INT-4208

Add AsyncRestTemplate support in HttpRequestExecutingMessageHandler and corresponding config/dsl support

pkuashi commented on spring-integration-http/src/main/java/org/springframework/integration/http/dsl/AsyncHttpMessageHandlerSpec.java:
```diff
@@ -0,0 +1,71 @@
+	AsyncHttpMessageHandlerSpec(Expression uriExpression, AsyncRestTemplate asyncRestTemplate) {
+		this.target = new AsyncHttpRequestExecutingMessageHandler(uriExpression, asyncRestTemplate);
+		this.target.setUriVariableExpressions(this.uriVariableExpressions);
+		this.target.setHeaderMapper(this.headerMapper);
```
You mean moving the assignments to the ctor of `BaseHttpMessageHandlerSpec`? I have tried that but faced one problem: I still want to leave the `this.target = new AsyncHttpRequestExecutingMessageHandler(uriExpression, asyncRestTemplate);` to the concrete subclass to avoid reflection solution then after moving `this.target.setUriVariableExpressions(this.uriVariableExpressions); this.target.setHeaderMapper(this.headerMapper);` statements to `BaseHttpMessageHandlerSpec` will throw error  due to the this.target is not initialized yet.

What you think?

artembilan replied:
???
How about this?
```
protected BaseHttpMessageHandlerSpec(E handler) {
	this.target = handler;
	this.target.setUriVariableExpressions(this.uriVariableExpressions);
	this.target.setHeaderMapper(this.headerMapper);
}
...
AsyncHttpMessageHandlerSpec(Expression uriExpression, AsyncRestTemplate asyncRestTemplate) {
	super(new AsyncHttpRequestExecutingMessageHandler(uriExpression, asyncRestTemplate));
	this.asyncRestTemplate = asyncRestTemplate;
}
...
HttpMessageHandlerSpec(Expression uriExpression, RestTemplate restTemplate) {
	super(new HttpRequestExecutingMessageHandler(uriExpression, restTemplate));
	this.restTemplate = restTemplate;
}
```

ShiliangHM replied:
good idea!

pkuashi replied:
Good idea!

pkuashi-official committed:
INT-4208: Support for AsycRestTemplate and thus noblocking IO

https://jira.spring.io/browse/INT-4076
https://jira.spring.io/browse/INT-4208

Add AsyncRestTemplate support in HttpRequestExecutingMessageHandler and corresponding config/dsl support

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/802061985cb8eff4c508aa246e59f5ee2c0364bc

Awesome work, @pkuashi !
Thank you very much for the contribution.

Looking forward for more!

pkuashi commented:
@artembilan Thanks for hosting my first PR :)
